### PR TITLE
Verify main vs extra JWT bearers differently

### PR DESCRIPTION
**Pull request by NickMeves**
When using the configured provider JWT Verifier, it makes
sense to use the provider `CreateSessionStateFromBearerToken`
method. For any extra JWT Issuers, they should use a generic
default verifier.

https://github.com/oauth2-proxy/oauth2-proxy/issues/592

## How Has This Been Tested?

Unit tests added for legacy/generic bearer IDToken -> SessionState.

TODO: deploy to a local kubernetes cluster and validate standard human OIDC with k8s service account tokens as an extra jwt issuer. UPDATE: DONE

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: oauthproxy.go [hunk 1]
```diff
@@ -1146,15 +1148,24 @@ func (p *OAuthProxy) GetJwtSession(req *http.Request) (*sessionsapi.SessionState
...
+			return p.provider.CreateSessionStateFromBearerToken(req.Context(), rawBearerToken, bearerToken)
+		}
+	}
 
+	// Otherwise, attempt to verify against the extra JWT issuers and use a more generic
+	// Bearer Token -> Session converter
+	for _, verifier := range p.extraJwtBearerVerifiers {
+		bearerToken, err := verifier.Verify(req.Context(), rawBearerToken)
 		if err != nil {
-			logger.Printf("failed to verify bearer token: %v", err)
```
Comment (NickMeves) [inline oauthproxy.go:1153]: I removed this and defaulted to the logging happening only if all verifiers failed.  My users have been confused with error messages when 1st or 2nd verifiers on the array fail and the 3rd extraJwt Verifier succeeded (seeing the `failed` words even though it succeeded was confusing)

File: providers/provider_default_test.go [hunk 2]
```diff
@@ -19,3 +23,35 @@ func TestRefresh(t *testing.T) {
...
+	rawIDToken, _ := jwt.NewWithClaims(jwt.SigningMethodRS256, minimalIDToken).SignedString(key)
+	idToken, err := verifier.Verify(context.Background(), rawIDToken)
+
+	session, err := (*ProviderData)(nil).CreateSessionStateFromBearerToken(context.Background(), rawIDToken, idToken)
+
+	assert.Equal(t, nil, err)
+	assert.Equal(t, rawIDToken, session.AccessToken)
+	assert.Equal(t, rawIDToken, session.IDToken)
+	assert.Equal(t, "", session.RefreshToken)
+	assert.Equal(t, "123456789", session.Email)
```
Comment (NickMeves) [inline providers/provider_default_test.go:54]: What is the overall architectural rationale for forcing a `session.Email` to a `claim.Subject` when an `email` claim is missing? And not using the `claim.Subject` at all in the session and setting both the `session.Email` and `session.User` to the Email?

https://github.com/oauth2-proxy/oauth2-proxy/blob/f7b28cb1d362f54d13379136c37f0edc1303a1b0/providers/provider_default.go#L155

I left the existing logic untouched -- but shouldn't `session.User` == `claim.Subject` (That's what my OIDC provider results in going the normal route).

https://github.com/oauth2-proxy/oauth2-proxy/blob/f7b28cb1d362f54d13379136c37f0edc1303a1b0/providers/provider_default.go#L165

And should we allow `session.Email` to be "" if the IDToken is missing the claim?  And adjust the email domains validator to accept it if its set to `*` (if set to anything else it would fail since it is missing a `@` character)

**Comment by NickMeves** [comment]
On further reflection, I think this is definitely a bug? https://github.com/oauth2-proxy/oauth2-proxy/blob/a17c48810fc2b5267be256cb4812584c14ccb9d1/providers/provider_default.go#L165

I had a user make a web app that used this proxy for use with CLI tools that did OIDC via PKSE and passed the IDToken as the bearer. He got the `X-Forwarded-User` as the `email` claim, vs all the other normal webapps made with our OneLogin OIDC provider had the `subject` claim as the `X-Forwarded-User`.

I was very confused at the time, but this is definitely why.

File: providers/provider_default.go [hunk 3]
```diff
@@ -162,14 +162,13 @@ func (p *ProviderData) CreateSessionStateFromBearerToken(ctx context.Context, ra
 
 	newSession := &sessions.SessionState{
 		Email:             claims.Email,
-		User:              claims.Email,
+		User:              claims.Subject,
```
Comment (NickMeves) [inline providers/provider_default.go:165]: Let me know if you think we should move this to a different PR (since this required changing a test in `oauth2proxy_test.go` potentially changing expected behavior by users).

This now aligns this to how the OIDC `CreateSessionStateFromBearerToken` provider would've extracted claims to sessions.

Comment (JoelSpeed) [reply providers/provider_default.go:165]: Agreed, this is a bug, it's `claims.Subject` in the OIDC provider so happy to leave it like this for now. Since 499 hasn't made it into a release yet, no need to consider this breaking

File: providers/provider_default_test.go (hunk 2 above)
Comment (JoelSpeed) [reply providers/provider_default_test.go:54]: User ID claim was added so that we could make sure that a particular claim was added, to be considered the user, but the work hasn't been completed so it's only half way through

As for falling back, I'm not sure why we are doing that, i think we currently require all sessions to have a user and email field that's not empty, so I guess that's why? I think I need to do some archaeology on this

**Comment by NickMeves** [comment]
I still owe you an interactive test on this, I haven't dockerized this branch yet to test how it all works deployed as a sidecar in a local kubernetes environment that is setup for OIDC projected service account tokens.

**Comment by NickMeves** [comment]
Should I make the `default_provider` aware of the `--user-id-claim` flag when mapping claims to a session.Email just the same as the OIDC provider?

File: providers/provider_default.go (hunk 3 above)
Comment (NickMeves) [reply providers/provider_default.go:165]: So if I recall correctly, 499 mostly moved this code verbatim from oauth2proxy.go to the default_provider.

The previous versions also had `User: claims.Email` (hence my question about this fix needing a note that this is breaking)

**Comment by NickMeves** [comment]
@JoelSpeed Let me know what you find out as far as the history of the forced `session.Email` and not allowing it to be "".

This commit is my take on allowing it: https://github.com/oauth2-proxy/oauth2-proxy/pull/596/commits/f2bf5a39b19e19cf8881b34205153f18c8ad84df

I can ditch it if needed (or cherry pick it to a separate PR to keep this PR purely about the bearer bugs).

**Comment by NickMeves** [comment]
> I still owe you an interactive test on this, I haven't dockerized this branch yet to test how it all works deployed as a sidecar in a local kubernetes environment that is setup for OIDC projected service account tokens.

This is done, looks good in a kubernetes environment allowing both OneLogin IDTokens (configured provider) OR kubernetes projected service account tokens (extra bearer JWT):

```
[2020/06/03 17:15:09] [oauthproxy.go:283] Skipping JWT tokens from configured OIDC issuer: "https://greenhouse.onelogin.com/oidc/2"
[2020/06/03 17:15:09] [oauthproxy.go:285] Skipping JWT tokens from extra JWT issuer: "http://oidc.default.svc.cluster.local=sample-app"
...
192.168.144.128 - nick.meves@greenhouse.io [2020/06/03 17:21:27] sample-app.local GET 127.0.0.1:5000 "/sso" HTTP/1.1 "Ruby" 200 1336 0.003
192.168.144.128 - system:serviceaccount:default:dummy [2020/06/03 17:23:46] sample-app.local GET 127.0.0.1:5000 "/sso" HTTP/1.1 "Ruby" 200 1174 0.005
```

**Comment by NickMeves** [comment]
I had the `/sso` endpoint just spit back the `X-Forwarded-*` auth headers, you can see the behavior allowing empty session.Email in the service account token output (the experimental commit we can remove from this this PR if desired):

OneLogin OIDC Human IDToken:
```
{
  "data": {
     "username":  "[REDACTED]",
     "prefered_username": "nick.meves@greenhouse.io",
     "email":  "nick.meves@greenhouse.io"
  }
}
```

Kubernetes Projected Service Account IDToken:
```
{
  "data": {
    "username": "system:serviceaccount:default:dummy",
    "prefered_username": null,
    "email": null
  }
}
```

**Comment by JoelSpeed** [comment]
So I did some archaeology today, my conclusion is that, if the scope contained `email`, then the response and the userinfo endpoint should have the `email` value, but if `email` is not in the scope, then we don't need for it to fallback to anything I don't think

As far as I can tell, the OAuth2 Proxy doesn't require the email to be returned from the provider anywhere, so I think we can get away with not setting it. It's probably better that I rework this into #560 

This will likely be a breaking change, though, technically better than the existing behaviour.
I think we should probably look into the user-id-claim work and work out how we move forward with that.

**Comment by NickMeves** [comment]
> So I did some archaeology today, my conclusion is that, if the scope contained `email`, then the response and the userinfo endpoint should have the `email` value, but if `email` is not in the scope, then we don't need for it to fallback to anything I don't think
> 
> As far as I can tell, the OAuth2 Proxy doesn't require the email to be returned from the provider anywhere, so I think we can get away with not setting it. It's probably better that I rework this into #560
> 
> This will likely be a breaking change, though, technically better than the existing behaviour.
> I think we should probably look into the user-id-claim work and work out how we move forward with that.

I can remove that commit, isn't the end of the world to have a `session.Email` that isn't an email in cases where the subject claim isn't an email (with it as blank or a non-email from the subject -- both routes require `--emails-domains=*` to validate appropriately).

I'm not sure this is best handled in the #560 -- its that weird middle ground for how do we agnostically treat extra JWT bearers that aren't from the main provider in a generic way.

IE my case, I like my OIDC provider settings have the user-id claim to `email` like the default, but I don't want to change that to something just to appease compatibility with whatever extra JWT Bearer Verifiers I add since their IDToken structures could feasibly be all over the place (subject/audience is likely the only thing we can rely on it having).

File: providers/provider_default_test.go [hunk 4]
```diff
@@ -47,3 +52,35 @@ func TestAcrValuesConfigured(t *testing.T) {
...
+		&oidc.Config{ClientID: "asdf1234"},
+	)
+
+	key, _ := rsa.GenerateKey(rand.Reader, 2048)
+	rawIDToken, _ := jwt.NewWithClaims(jwt.SigningMethodRS256, minimalIDToken).SignedString(key)
+	idToken, _ := verifier.Verify(context.Background(), rawIDToken)
+
+	session, err := (*ProviderData)(nil).CreateSessionStateFromBearerToken(context.Background(), rawIDToken, idToken)
+
+	assert.Equal(t, nil, err)
```
Comment (NickMeves) [inline providers/provider_default_test.go:79]:
```suggestion
	assert.NoError(t, err)
```

File: providers/provider_default_test.go [hunk 5]
```diff
@@ -47,3 +52,39 @@ func TestAcrValuesConfigured(t *testing.T) {
...
+		fakeKeySetStub{},
+		&oidc.Config{ClientID: "asdf1234"},
+	)
+
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	assert.NoError(t, err)
+	rawIDToken, err := jwt.NewWithClaims(jwt.SigningMethodRS256, minimalIDToken).SignedString(key)
+	assert.NoError(t, err)
+	// Pass to a dummy Verifier to get an oidc.IDToken from the rawIDToken for our actual test below
+	idToken, err := verifier.Verify(context.Background(), rawIDToken)
```
Comment (NickMeves) [inline providers/provider_default_test.go:78]: There's got to be a cleaner way to get an `oidc.IDToken` from a rawIDToken string...? Getting it through a dummy `oidc.NewVerifier` worked, but just feels sloppy to me -- but nothing stood out to me when I tried to dig through `oidc` 🤷

**Comment by NickMeves** [comment]
> @NickMeves Can you just confirm this is good to go from your perspective, I believe it is and have no further comments
> 
> @steakunderscore I'd appreciate if you could also review this one before merging

Yup, ready to go.  I ran some manual tests in a local kubernetes environment and it looked good.

Only area I had a lingering question is the new unit test I added that feels a little sloppy.

File: providers/provider_default_test.go (hunk 5 above)
Comment (JoelSpeed) [reply providers/provider_default_test.go:78]: It is just a struct at the end of the day, you could theoretically construct it, but then you can't set the private methods. I think what you've done is probably the best we can do sadly

**Comment by NickMeves** [comment]
Give me the heads up when we're ready and I'll rebase before the merge.  I think I keep dismissing your reviews when I push up rebases 😄

**Comment by JoelSpeed** [comment]
I think let's just get this merged, can you rebase?
